src/js/plugin.js

```diff
@@ -0,0 +1,445 @@
+  static isBasic(plugin) {
+    const p = (typeof plugin === 'string') ? getPlugin(plugin) : plugin;
+
+    return typeof p === 'function' && !Plugin.prototype.isPrototypeOf(p.prototype);
+  }
+
+  /**
+   * Register a Video.js plugin.
+   *
+   * @param  {string} name
+   *         The name of the plugin to be registered. Must be a string and
+   *         must not match an existing plugin or a method on the `Player`
+   *         prototype.
+   *
+   * @param  {Function} plugin
+   *         A sub-class of `Plugin` or a function for basic plugins.
+   *
+   * @return {Function}
+   *         For advanced plugins, a factory function for that plugin. For
+   *         basic plugins, a wrapper function that initializes the plugin.
+   */
+  static registerPlugin(name, plugin) {
+    if (typeof name !== 'string') {
+      throw new Error(`Illegal plugin name, "${name}", must be a string, was ${typeof name}.`);
+    }
+
+    if (pluginExists(name) || Player.prototype.hasOwnProperty(name)) {
+      throw new Error(`Illegal plugin name, "${name}", already exists.`);
+    }
+
+    if (typeof plugin !== 'function') {
+      throw new Error(`Illegal plugin for "${name}", must be a function, was ${typeof plugin}.`);
+    }
+
+    pluginStorage[name] = plugin;
+
+    // Add a player prototype method for all sub-classed plugins (but not for
+    // the base Plugin class).
+    if (name !== BASE_PLUGIN_NAME) {
+      if (Plugin.isBasic(plugin)) {
+        Player.prototype[name] = createBasicPlugin(name, plugin);
+      } else {
+        Player.prototype[name] = createPluginFactory(name, plugin);
+      }
+    }
+
+    return plugin;
+  }
+
+  /**
+   * De-register a Video.js plugin.
+   *
+   * @param  {string} name
+   *         The name of the plugin to be deregistered.
+   */
+  static deregisterPlugin(name) {
+    if (name === BASE_PLUGIN_NAME) {
+      throw new Error('Cannot de-register base plugin.');
+    }
+    if (pluginExists(name)) {
+      delete pluginStorage[name];
+      delete Player.prototype[name];
+    }
+  }
+
+  /**
+   * Gets an object containing multiple Video.js plugins.
+   *
+   * @param  {Array} [names]
+   *         If provided, should be an array of plugin names. Defaults to _all_
+   *         plugin names.
+   *
+   * @return {Object|undefined}
+   *         An object containing plugin(s) associated with their name(s) or
+   *         `undefined` if no matching plugins exist).
+   */
+  static getPlugins(names = Object.keys(pluginStorage)) {
+    let result;
+
+    names.forEach(name => {
+      const plugin = getPlugin(name);
+
+      if (plugin) {
+        result = result || {};
+        result[name] = plugin;
+      }
+    });
+
+    return result;
+  }
+
+  /**
+   * Gets a plugin by name if it exists.
+   *
+   * @param  {string} name
+   *         The name of a plugin.
+   *
+   * @return {Function|undefined}
+   *         The plugin (or `undefined`).
+   */
+  static getPlugin(name) {
+    return getPlugin(name);
```

Comment:
the names being the same makes it a bit confusing. Took me a second to realize it wouldn't be an infinite loop.